File: 2-Basic-Mathematical-Knowledge-for-Algorithms/009_Brute_Force_2/java/Main.java
import java.util.Scanner;

public class Main {
    static int n;
    static long s;
    static int[] a;
    static boolean[] possible1;
    static boolean found = false;

    // 前半部分の和を列挙し、possible1配列に記録
    static void generate1(int idx, int endIdx, long currentSum) {
        if (currentSum > s) { // 枝刈り
            return;
        }
        if (idx == endIdx) {
            possible1[(int)currentSum] = true;
            return;
        }
        generate1(idx + 1, endIdx, currentSum); // a[idx] を選ばない
        generate1(idx + 1, endIdx, currentSum + a[idx]); // a[idx] を選ぶ
    }

    // 後半部分の和を列挙し、答えを探索
    static void findSolution(int idx, int endIdx, long currentSum) {
        if (found) return; // 既に発見済みなら即終了
        if (currentSum > s) { // 枝刈り
            return;
        }
        if (idx == endIdx) {
            long need = s - currentSum;
            if (need >= 0 && need <= s && possible1[(int)need]) {
                found = true;
            }
            return;
        }
        findSolution(idx + 1, endIdx, currentSum); // a[idx] を選ばない
        findSolution(idx + 1, endIdx, currentSum + a[idx]); // a[idx] を選ぶ
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        
        // 1. 入力処理
        n = sc.nextInt();
        s = sc.nextLong();
        
        a = new int[n];
        for (int i = 0; i < n; i++) {
            a[i] = sc.nextInt();
        }
        
        possible1 = new boolean[(int)s + 1];
        
        // 2. 前半と後半に分けて処理
        int half1End = n / 2;
        
        generate1(0, half1End, 0);
        findSolution(half1End, n, 0);
        
        // 3. 結果出力
        if (found) {
            System.out.println("Yes");
        } else {
            System.out.println("No");
        }
        
        sc.close();
    }
}

File: 2-Basic-Mathematical-Knowledge-for-Algorithms/009_Brute_Force_2/c/main.c
#include <stdio.h>
#include <stdbool.h>

int n;
long long s;
int a[60];
bool possible1[10001];
bool found = false;

// 前半部分の和を列挙し、possible1配列に記録
void generate1(int idx, int end_idx, long long current_sum) {
    if (current_sum > s) { // 枝刈り
        return;
    }
    if (idx == end_idx) {
        possible1[current_sum] = true;
        return;
    }
    generate1(idx + 1, end_idx, current_sum); // a[idx] を選ばない
    generate1(idx + 1, end_idx, current_sum + a[idx]); // a[idx] を選ぶ
}

// 後半部分の和を列挙し、答えを探索
void find_solution(int idx, int end_idx, long long current_sum) {
    if (found) return; // 既に発見済みなら即終了
    if (current_sum > s) { // 枝刈り
        return;
    }
    if (idx == end_idx) {
        long long need = s - current_sum;
        if (need >= 0 && need <= s && possible1[need]) {
            found = true;
        }
        return;
    }
    find_solution(idx + 1, end_idx, current_sum); // a[idx] を選ばない
    find_solution(idx + 1, end_idx, current_sum + a[idx]); // a[idx] を選ぶ
}

int main() {
    // 1. 入力処理
    scanf("%d %lld", &n, &s);
    for (int i = 0; i < n; i++) {
        scanf("%d", &a[i]);
    }
    
    // 初期化
    for (int i = 0; i <= s; i++) {
        possible1[i] = false;
    }
    
    // 2. 前半と後半に分けて処理
    int half1_end = n / 2;
    
    generate1(0, half1_end, 0);
    find_solution(half1_end, n, 0);
    
    // 3. 結果出力
    if (found) {
        printf("Yes\n");
    } else {
        printf("No\n");
    }
    
    return 0;
}